src/security/KeyData.cc

```diff
@@ -94,51 +133,40 @@ Security::KeyData::loadX509ChainFromFile()
-        OPENSSL_free(nameStr);
-    } else
-#endif
-    {
-        debugs(83, DBG_PARSE_NOTE(3), "Using certificate chain in " << certFile);
-        // and add to the chain any other certificate exist in the file
-        CertPointer latestCert = cert;
-
-        while (auto ca = PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr)) {
-            // get Issuer name of the cert for debug display
-            char *nameStr = X509_NAME_oneline(X509_get_subject_name(ca), nullptr, 0);
-
-#if TLS_CHAIN_NO_SELFSIGNED // ignore self-signed certs in the chain
-            // self-signed certificates are not valid in a sent chain
-            if (X509_check_issued(ca, ca) == X509_V_OK) {
-                debugs(83, DBG_PARSE_NOTE(2), "CA " << nameStr << " is self-signed, will not be chained: " << nameStr);
-                OPENSSL_free(nameStr);
-                continue;
-            }
-#endif
-            // checks that the chained certs are actually part of a chain for validating cert
-            const auto checkCode = X509_check_issued(ca, latestCert.get());
-            if (checkCode == X509_V_OK) {
-                debugs(83, DBG_PARSE_NOTE(3), "Adding issuer CA: " << nameStr);
-                // OpenSSL API requires that we order certificates such that the
-                // chain can be appended directly into the on-wire traffic.
-                latestCert = CertPointer(ca);
-                chain.emplace_front(latestCert);
-            } else {
-                debugs(83, DBG_PARSE_NOTE(2), certFile << ": Ignoring non-issuer CA " << nameStr << ": " << X509_verify_cert_error_string(checkCode) << " (" << checkCode << ")");
-            }
-            OPENSSL_free(nameStr);
-        }
+    while (CertPointer ca = CertPointer(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr))) {
+        verifyChainCert(ca, latestCert, chain);
     }
 
 #elif USE_GNUTLS
-    // XXX: implement chain loading
-    debugs(83, 2, "Loading certificate chain from PEM files not implemented in this Squid.");
+    const char *certFilename = certFile.c_str();
+    gnutls_datum_t data;
+    Security::ErrorCode x = gnutls_load_file(certFilename, &data);
+    if (x != GNUTLS_E_SUCCESS) {
+        debugs(83, DBG_IMPORTANT, "ERROR: unable to load chain file '" << certFile << "': " << ErrorString(x));
+        return;
+    }
+
+    unsigned int listSz = 0;
+    gnutls_x509_crt_t *certChain;
+    x = gnutls_x509_crt_list_import2(&certChain, &listSz, &data, GNUTLS_X509_FMT_PEM, 0);
+    if (x != GNUTLS_E_SUCCESS) {
+        debugs(83, DBG_IMPORTANT, "ERROR: unable to import chain file '" << certFile << "': " << ErrorString(x));
+        return;
+    }
+
+    for (unsigned int i = 0; i < listSz ; ++i) {
+        CertPointer ca = Security::CertPointer(certChain[i], [](gnutls_x509_crt_t p) {
+            debugs(83, 5, "gnutls_x509_crt_deinit cert=" << (void*)p);
+            gnutls_x509_crt_deinit(p);
```

Comment:
This is the TLS "certificate chain" from root CA down to issued certificate. Its library API representation happens to be an array of raw-pointers to cert objects.

We are passing ownership of each raw-pointer in turn to a CertPointer object and supplying a lambda to be used as destructor. How and when that object is destructed is up to its new owner. We only need to free the original array memory - which happens following the loop.

Reply:
Thank you for confirming my understanding. My suggestion stands.

> We are passing ownership ...

FYI: The current code does not express any change of ownership. A new owner is indeed created while the old owner stays intact. The intent may be to "pass ownership", but the code does not express that intent well (and, to some readers, at all).

The root of this problem is in the ownership-ignorant C API -- we probably cannot fully address the underlying issue here -- but the suggested renaming may alleviate some of the associated pain and clarify the intent. Without dereferencing each item, it is usually not possible to free the underlying memory of a _chain_ of items, but it is usually possible to free the underlying memory of an array of items.

If you are adamantly against this adjustment, I can live with the current code, but please add a `does not touch the already-freed individual certificates` comment to the dangerous-looking gnutls_free() call.

Reply:
Lets start this thread again...

> Deleting chained certificates while we are still iterating the chain feels dangerous to me, even if it works.

Certificates are not being deleted here. The raw-pointer stored in array "certChain" is being turned into a Pointer. The *_free() code line you see here is a statement of the lambda which runs when that Pointer destructs - however that destruct does not happen until the socket closes or similar.

> Without dereferencing each item, it is usually not possible to free the underlying memory of a chain of items, but it is usually possible to free the underlying memory of an array of items.

The object being free'd after the loop is implemented as array ( int *[]  ). The local varible is named "certChain" because 'chain' is TLS terminology for the collection - not C/C++ terminology.